elements/pfe-datetime/pfe-datetime.ts

```diff
@@ -0,0 +1,215 @@
+
+import style from './pfe-datetime.scss';
+
+function dash(str: string): string {
+  return str
+    .replace(/[\w]([A-Z])/g, match => `${match[0]}-${match[1]}`)
+    .toLowerCase();
+}
+
+const INTL_CONFIGS = Object.entries({
+  weekday: {
+    short: 'short',
+    long: 'long',
+  },
+  day: {
+    'numeric': 'numeric',
+    '2-digit': '2-digit',
+  },
+  month: {
+    short: 'short',
+    long: 'long',
+  },
+  year: {
+    'numeric': 'numeric',
+    '2-digit': '2-digit',
+  },
+  hour: {
+    'numeric': 'numeric',
+    '2-digit': '2-digit',
+  },
+  minute: {
+    'numeric': 'numeric',
+    '2-digit': '2-digit',
+  },
+  second: {
+    'numeric': 'numeric',
+    '2-digit': '2-digit',
+  },
+  timeZoneName: {
+    short: 'short',
+    long: 'long',
+  },
+}) as [keyof Required<Intl.DateTimeFormatOptions>, object][];
+
+/**
+ * @attr {'narrow'|'short'|'long'} weekday
+ * @attr {'numeric'|'2-digit'|'narrow'|'short'|'long'} month
+ * @attr {'numeric'|'2-digit'} day
+ * @attr {'numeric'|'2-digit'} year
+ * @attr {'numeric'|'2-digit'} hour
+ * @attr {'numeric'|'2-digit'} minute
+ * @attr {'numeric'|'2-digit'} second
+ */
+@customElement('pfe-datetime') @pfelement()
+export class PfeDatetime extends LitElement {
+  static readonly styles = [style]
+
+  /**
+   * The options for type are:
+   * - `local`: Shows a formatted time for the indicated locale if provided
+   * - `relative`: Shows a relative time (1 hour ago, 2 hours until)
+   */
+  @property({ reflect: true }) type: 'local'|'relative' = 'local';
+
+  /** The value of this should be the same timestamp that you add to the light DOM. */
+  @observed
+  @property({ reflect: true }) datetime = '';
+
+  /**
+   * A unix timestamp that will be converted for use in displaying the appropriate date.
+   * You would not use both datetime and timestamp, and the last updated will take precedence.
+   */
+  @observed
+  @property({ type: Number, reflect: true }) timestamp?: number;
+
+  /** Time Zone */
+  @property({ reflect: true, attribute: 'time-zone' }) timeZone?: string;
+
+  @state() private _datetime?: number;
+
+  @state() private _timestamp?: number;
+
+  private logger = new Logger(this);
+
+  private get _dateTimeType() {
+    return this.type || 'local';
+  }
+
+  render() {
+    const formattedDate = this._getTypeString();
+    return html`
+      <span>${formattedDate}</span>
+    `;
+  }
+
+  protected _datetimeChanged(_oldVal?: string, newVal?: string) {
+    if (!newVal || !Date.parse(newVal))
+      return;
+
+
+    if (Date.parse(newVal) && this._datetime === Date.parse(newVal))
+      return;
+
+
+    this.setDate(Date.parse(newVal));
+  }
+
+  protected _timestampChanged(_oldVal?: number, newVal?: number) {
+    if (newVal == null || this._timestamp === newVal) {
+      this.logger.log('early return');
+      return;
+    }
+
+    this._timestamp = newVal;
+    this.setDate(new Date(newVal * 1000));
+  }
+
+  private _getOptions(): Intl.DateTimeFormatOptions {
```

Comment:
`Intl.DateTimeFormatOptions` that's a really useful feature of TS. 😀

Reply:
dom.lib.ts top reads of 2021